File: Image/Extensions/Geography/Geography.cs
using TahsinsLibrary.Geometry;
using System.Collections.Generic;
namespace TahsinsLibrary.Geography
{
    public sealed class Culture
    {
        public string name;
        public Color color;
        public Culture(string name, Color color)
        {
            this.name = name;
            this.color = color;
        }
    }
    public sealed class Biome
    {
        public string name;
        public Color color;
        public bool canPlaceCountry;
        public Biome(string name, Color color, bool canPlaceCountry)
        {
            this.name = name;
            this.color = color;
            this.canPlaceCountry = canPlaceCountry;
        }

    }
    public sealed class Settlement
    {
        public string name;
        public enum Status
        {
            Village, Town, City, Metropolis
        }
        public Status status;
        public Vector2D location;
        public Settlement(string name, Status status, Vector2D location)
        {
            this.name = name;
            this.status = status;
            this.location = location;
        }
    }
    public abstract class PoliticalEntity
    {
        protected Settlement capital;
        public string name;
        public PoliticalEntity(string name ,Settlement capital)
        {
            this.name = name;
            this.capital = capital;
        }
    }
    public sealed class Province : PoliticalEntity
    {
        public State owner;
        private List<Settlement> settlements;

        public Province(string name, Settlement capital, List<Settlement> settlements) : base(name, capital)
        {
            this.settlements = settlements;
            this.name = name;
            if(!settlements.Contains(capital))
            {
                settlements.Add(capital);
            }
            this.capital = capital;
        }

        public void AddSettlement(Settlement settlement)
        {
            if(!settlements.Contains(settlement))
            {
                settlements.Add(settlement);
            }
        }
        public bool HasSettlement(Settlement settlement) => settlements.Contains(settlement);
    }
    public sealed class State : PoliticalEntity
    {
        private List<Province> provinces;

        public State(string name, Settlement capital, List<Province> provinces) : base(name, capital)
        {
            this.name = name;
            this.provinces = provinces;
            bool found = false;
            foreach(Province province in provinces)
            {
                if(province.HasSettlement(capital))
                {
                    found = true;
                    break;
                }
            }
            if(!found)
            {
                provinces.Add(new Province("Capital Proivnce",capital,new List<Settlement>{capital}));
            }
            this.capital = capital;
        }

        public void AddProvince(Province province)
        {
            if(!provinces.Contains(province))
            {
                provinces.Add(province);
            }
        }
        public Province RemoveProvince(Province province)
        {
            if(provinces.Contains(province))
            {
                provinces.Remove(province);
                return province;
            }
            else return null;
        }
        public bool HasProvince(Province province) => provinces.Contains(province);
        public bool HasSettlement(Settlement settlement)
        {
            foreach(Province province in provinces)
            {
                if(province.HasSettlement(settlement)) return true;
            }
            return false;
        }
    }
}

File: Image/Extensions/Geography/Geography.cpp
#include "Extensions/Geometry/Geometry.cpp"
#include "ImgLibBase.cpp"
#include <vector>
#include <string>
namespace Tahsin
{
    class Culture
    {
    public:
        string name;
        Color color;
        Culture(string name, Color color)
        {
            this->color = color;
            this->name = name;
        }
        ~Culture() {}
    };
    class Biome
    {
    public:
        string name;
        Color color;
        bool canPlaceCountry;
        Biome(string name, Color color, bool canPlaceCountry)
        {
            this->name = name;
            this->color = color;
            this->canPlaceCountry = canPlaceCountry;
        }
    };
    class Settlement
    {
    public:
        string name;
        enum Status
        {
            Village,
            Town,
            City,
            Metropolis
        };
        Status status;
        Vector2D location;
        Settlement() {}
        Settlement(string name, Status status, Vector2D location)
        {
            this->name = name;
            this->status = status;
            this->location = location;
        }
        ~Settlement() {}
    };
    class PoliticalEntity
    {
    protected:
        Settlement *capital;

    public:
        string name;
        PoliticalEntity() {}
    };
    class Province : PoliticalEntity
    {
    public:
        State *owner;
        Province() {}
        Province(string name, Settlement *capital, vector<Settlement *> settlements)
        {
            this->name = name;
            if (!count(settlements.begin(), settlements.end(), capital))
            {
                settlements.push_back(capital);
            }
            this->capital = capital;
            this->settlements = settlements;
        }
        bool HasSettlement(Settlement *settlement)
        {
            for (Settlement *s : settlements)
            {
                if (s == settlement)
                    return true;
            }
            return false;
        }

    private:
        vector<Settlement *> settlements;
    };
    class State : PoliticalEntity
    {
    private:
        vector<Province *> provinces;

    public:
        State() {}
        bool HasSettlement(Settlement *settlement)
        {
            for (Province *p : provinces)
            {
                if (p->HasSettlement(settlement))
                    return true;
            }
            return false;
        }
        bool HasProvince(Province *province)
        {
            for (Province *p : provinces)
            {
                if (p == province)
                    return true;
            }
            return false;
        }
        
    };
}